internal/charmstore/search.go

```
@@ -0,0 +1,285 @@
+	fields := map[string]float64{
+		"CharmMeta.Description": 3,
+	}
+	if sp.AutoComplete {
+		fields["CharmMeta.Name.ngrams"] = 10
+	} else {
+		fields["CharmMeta.Name"] = 10
+	}
+	return fields
+}
+
+// encodeFields takes a map of field name to weight and builds a slice of strings
+// representing those weighted fields for a MultiMatchQuery.
+func encodeFields(fields map[string]float64) []string {
+	fs := make([]string, 0, len(fields))
+	for k, v := range fields {
+		fs = append(fs, elasticsearch.BoostField(k, v))
+	}
+	return fs
+}
+
+// createSearchDSL builds an elasticsearch query from the query parameters.
+// http://www.elasticsearch.org/guide/en/elasticsearch/reference/current/query-dsl.html
+func createSearchDSL(sp SearchParams) string {
+	qdsl := elasticsearch.QueryDSL{
+		Size: sp.Limit,
+	}
+
+	// Full text search
+	var q elasticsearch.Query
+	if sp.Text == "" {
+		q = elasticsearch.MatchAllQuery{}
+	} else {
+		q = elasticsearch.MultiMatchQuery{
+			Query:  sp.Text,
+			Fields: encodeFields(queryFields(sp)),
+		}
+	}
+
+	// Attenuation function
+	q = elasticsearch.FunctionScoreQuery{
+		Query: q,
+		Functions: []elasticsearch.Function{
+			{
+				Function: "linear",
+				Field:    "UploadTime",
+				Scale:    "365d",
+			},
+		},
+	}
+
+	// Filters
+	qdsl.Query = elasticsearch.FilteredQuery{
+		Query:  q,
+		Filter: createFilters(sp.Filters),
+	}
+
+	bytes, err := json.Marshal(qdsl)
+	if err != nil {
+		panic(err)
+	}
+	return string(bytes)
+}
+
+// createFilters converts the filters requested with the serch API into
+// filters in the elasticsearch query DSL. Please see http://tinyurl.com/qzobc69
+// for details of how filters are specified in the API. For each key in f a filter is
+// created that matches any one of the set of values specified for that key.
+// The created filter will only match when at least one of the requested values
+// matches for all of the requested keys. createFilters will silently skip over any
```

// Any filter names that are not defined in the filters map will be silently skipped.
(to avoid the lower-case at start of sentence awkwardness) ?